vac/raw/decentralized-messaging-ethereum.md

```diff
@@ -0,0 +1,737 @@
+return (gamma, control, forward, I_sender, I_me)
+```
+
+#### process-create
+This function is called by the sender and each of the receivers of the *create* control message. 
+First, it records the information from the create message in the `gamma.history+`, 
+which is used to track group membership changes. Then, it proceeds to call `process-seed`.
+
+```
+op = (”create”, sender, seq, IDs)
+gamma$.history = gamma.history + {op}
+return (process-seed(gamma, sender, seq, dmsg))
+```
+
+#### process-ack
+This function is called by those group members once they receive an ack message. 
+In `process-ack`, `ackID` and `ackSeq` are the sender and 
+sequence number of the acknowledged message. 
+Firstly, if the acknowledged message is a group membership operation, 
+it records the acknowledgement in `gamma.history`. 
+
+Following this, the function retrieves the relevant member secret from `gamma.memberSecret`,
+which was previously obtained from the seed secret contained in the acknowledged message.
+
+Finally, it updates the ratchet for the sender of the *ack* and 
+returns the resulting update secret.
+
+```
+if (ackID, ackSeq) was a create / add / remove then
+op = ("ack", sender, seq, ackID, ackSeq)
+gamma$.history = gamma.history + {op}`
+s = gamma.memberSecret[ackID, ackSeq, sender]
+gamma$.memberSecret[ackID, ackSeq, sender] = empty_string
+if (s = empty_string) & (dmsg = empty_string) then return (gamma, empty_string, empty_string, empty_string, empty_string)
+if (s = empty_string) then (gamma, s) = decrypt-from(gamma, sender, dmsg)
+(gamma, I) = update-ratchet(gamma, sender, s)
+return (gamma, empty_string, empty_string, I, empty_string)
+```
+
+The HKDF function MUST follow RFC 5869 using the hash function SHA256.
+
+
+### Post-compromise security updates and group member removal
+
+The functions `update` and `remove` share similarities with `create`: 
+they both call the function `generate-seed` to encrypt a new seed secret for each group member. 
+The distinction lies in the determination of the group members using `member-view`. 
+In the case of `remove`, the user being removed is excluded from the recipients of the seed secret. 
+Additionally, the control message they construct is designated with type *update* or *remove* respectively.
+
+Likewise, `process-update` and `process-remove` are akin to `process-create`. 
+The function `process-update` skips the update of gamma.history, 
+whereas `process-remove` includes a removal operation in the history.
+
+#### update
+```
+control = ("update", ++gamma.mySeq, empty_string)
+recipients = member-view(gamma, gamma.myId) - {gamma.myId}
+(gamma, dmsgs) = generate-seed(gamma, recipients)
+(gamma, _, _, I , _) = process-update(gamma, gamma.myId, gamma.mySeq, empty_string, empty_string)
+return (gamma, control, dmsgs, I)
+```
+
+#### remove
+```
+control = ("remove", ++gamma.mySeq, empty)
+recipients = member-view(gamma, gamma.myId) - {ID, gamma.myId}
```

Comment:
Does this mean that anyone can remove a user from a group? Will there be any restrictions on this?